### ISSUE TITLE
[MPMD-163] PMD issues are reported multiple times

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-pmd-plugin#512]
**[Mirko Friedenhagen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mfriedenhagen)** opened **[MPMD-163](https://issues.apache.org/jira/browse/MPMD-163?redirect=false)** and commented

After updating maven-pmd-plugin from 2.7.1 to 3.0, in some of our projects the number of pmd violations exploded. Some are valid (new rules), however I have a project where the same violation will show up more than 70 times!
See below for an example of this dramatic outcome:

```
[mifr@host SomeService]$ grep '^<violation' target/pmd.xml | wc
    3682   44655 1239557
[mifr@host SomeService]$ grep '^<violation' target/pmd.xml | sort | uniq | wc
      92    1120   30832
```

The number of source files in this project in total is 96, so a wild guess would be, that the plugin readds already existing violations (the first (unique) violation occurs once, the second 4 times and so on). I will try t take a look at the source code or to get a working testcase.

---

**Affects:** 3.0

**Attachments:**
- [MPMD-163-against-tag-3.0.patch](https://issues.apache.org/jira/secure/attachment/12717265/MPMD-163-against-tag-3.0.patch) (_3.05 kB_)
- [my-app.zip](https://issues.apache.org/jira/secure/attachment/12717248/my-app.zip) (_2.08 kB_)

2 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Reijo Tomperi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dvice)** commented

I see this problem also. You can see the duplicates if you create a project with two files in the same package A.java and B.java. If there is a violation in B.java, then it is reported twice. If the violation is in A.java, it gets reported only once. If you have A, B and C. Then violations in C are reported 3 times etc. Also files in other folders can affect this (but that gets more complicated to follow). I tested this with this file:

import java.util.ArrayList;
public class Foo
{
public Foo(final ArrayList\<String> foo)
{
}
}

Also note that I don't see this problem when using standalone PMD, so this seems to be issue in the PMD plugin itself. I'm using Apache Maven 2.2.1.

**Comment by jira-importer** [comment]
**[Reijo Tomperi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dvice)** commented

I attached my-app.zip, which is a sample project where error occurs.

It reports these two violations for othe same file:
Avoid unused constructor parameters such as 'foo'.	4
Avoid unused constructor parameters such as 'foo'.	4

It should report this violation only once. If you remove App.java or rename the class (and file) e.g. to Gapp, then the duplicate goes away.

This also causes outofmemory error in our project (requires more than 4 GB of memory because the report size grows so large. This renders the new version not only unusable to us, but we had to fall back to previous version (which was especially difficult, because the two versions use different paths in resultset configuration file.

And to make things even worse, maven 2 uses the latest version if you use "mvn pmd:pmd", even if you have specified it to use some other version in the pom.xml. Only way to use different version is to use command 'mvn org.apache.maven.plugins:maven-pmd-plugin:2.4:pmd'.

**Comment by jira-importer** [comment]
**[Mirko Friedenhagen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mfriedenhagen)** commented

When running the unit-tests, this shows as well:

```
Running org.apache.maven.plugin.pmd.PmdViolationCheckMojoTest
[INFO] You have 17 warnings. For more details see:/Users/mifr/workspace/foss/maven-pmd-plugin-3.0/target/test/unit/default-configuration/target/pmd.xml
[INFO] PMD Warning: def.configuration.App:80 Rule:UselessParentheses Priority:4 Useless parentheses..
[INFO] PMD Warning: def/configuration/App2.java:22 Rule:UnusedImports Priority:4 Avoid unused imports such as 'java.io.File'.
[INFO] PMD Warning: def/configuration/App2.java:22 Rule:UnusedImports Priority:4 Avoid unused imports such as 'java.io.File'.
[INFO] PMD Warning: def.configuration.AppSample:58 Rule:UselessParentheses Priority:4 Useless parentheses..
[INFO] PMD Warning: def.configuration.AppSample:58 Rule:UselessParentheses Priority:4 Useless parentheses..
[INFO] PMD Warning: def.configuration.AppSample:58 Rule:UselessParentheses Priority:4 Useless parentheses..
[INFO] PMD Failure: def.configuration.App:31 Rule:UnusedPrivateField Priority:3 Avoid unused private fields such as 'unusedVar2'..
[INFO] PMD Failure: def.configuration.App:56 Rule:EmptyCatchBlock Priority:3 Avoid empty catch blocks.
[INFO] PMD Failure: def.configuration.AppSample:27 Rule:UnusedPrivateField Priority:3 Avoid unused private fields such as 'unusedVar'..
[INFO] PMD Failure: def.configuration.AppSample:27 Rule:UnusedPrivateField Priority:3 Avoid unused private fields such as 'unusedVar'..
[INFO] PMD Failure: def.configuration.AppSample:27 Rule:UnusedPrivateField Priority:3 Avoid unused private fields such as 'unusedVar'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedFormalParameter Priority:3 Avoid unused method parameters such as 'unusedParam'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedFormalParameter Priority:3 Avoid unused method parameters such as 'unusedParam'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedFormalParameter Priority:3 Avoid unused method parameters such as 'unusedParam'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'unusedMethod(String)'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'unusedMethod(String)'..
[INFO] PMD Failure: def.configuration.AppSample:45 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'unusedMethod(String)'..
Caught Expected Message: You have 11 PMD violations and 6 warnings. For more details see:/Users/mifr/workspace/foss/maven-pmd-plugin-3.0/target/test/unit/default-configuration/target/pmd.xml
```

**Comment by jira-importer** [comment]
**[Melloware](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elefkof)** commented

I just upgraded to 3.0 this weekend and I am seeing the exact same error.  Tons of duplicates reported repeatedly in the PMD output.

**Comment by jira-importer** [comment]
**[Mirko Friedenhagen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mfriedenhagen)** commented

Fix against https://svn.apache.org/repos/asf/maven/plugins/tags/maven-pmd-plugin-3.0, works against trunk as well, sorry for the svn:ignore.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

fixed http://svn.apache.org/r1447809
Thanks !

**Comment by jira-importer** [comment]
**[Florent PAILLARD](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=florentp)** commented

Oups! Same problem here... The number of error is growing exponentially, resulting in OutOfMemory errors...

I hope the 3.1 will be released soon! :)

Thanks guys for your work!

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

did you try 3.1-SNAPSHOT ?

**Comment by jira-importer** [comment]
**[Florent PAILLARD](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=florentp)** commented

Great! I haven't seen it, my local Maven proxy was not up-to-date.

Works perfectly with 3.1-SNAPSHOT.

Thanks Oliver
